src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +137,129 @@ private void logTargetPath() {
-                aprilTagCamera.calculateStandardDeviations()
-        );
+        final Pose2d estimatedOdometryPose = estimatedPoseAtObservationTime.plus(odometryPoseToObservationTimestampPoseTransform.inverse());
+        this.estimatedPose = estimatedOdometryPose.plus(calculatePoseAmbiguityAsTransform(estimatedPoseAtObservationTime, standardDeviations));
     }
 
-    private void putAprilTagsOnFieldWidget() {
-        for (Map.Entry<Integer, Pose3d> entry : FieldConstants.TAG_ID_TO_POSE.entrySet()) {
-            final Pose2d tagPose = entry.getValue().toPose2d();
-            field.getObject("Tag " + entry.getKey()).setPose(tagPose);
+    /**
+     * Checks if the buffer that stores the previous poses of the robot still has something stored at the target timestamp.
+     *
+     * @param timestamp the target timestamp to check
+     * @return whether the buffer contains a value at that timestamp or not
+     */
+    private boolean isVisionObservationTooOld(double timestamp) {
+        try {
+            final double oldestEstimatedRobotPoseTimestamp = previousOdometryPoses.getInternalBuffer().lastKey() - PoseEstimatorConstants.POSE_BUFFER_SIZE_SECONDS;
+            if (oldestEstimatedRobotPoseTimestamp > timestamp)
+                return true;
+        } catch (NoSuchElementException e) {
+            return true;
         }
+        return false;
+    }
+
+    /**
+     * Gets the estimated pose of the robot at the target timestamp.
+     *
+     * @param timestamp the target timestamp
+     * @return the robot's estimated pose at the timestamp
+     */
+    public Pose2d getPoseAtTimestamp(double timestamp) {
+        final Pose2d poseAtTimestamp = previousOdometryPoses.getSample(timestamp).orElse(new Pose2d());
+        final Transform2d odometryPoseToPoseAtTimestampTransform = new Transform2d(odometryPose, poseAtTimestamp);
+
+        return estimatedPose.plus(odometryPoseToPoseAtTimestampTransform);
+    }
+
+    /**
+     * Calculates the ambiguity of the estimated pose from the standard deviations of the camera and stores the values as a {@link Transform2d}.
+     *
+     * @param estimatedPoseAtObservationTime the pose estimate at the timestamp of the camera's observation
+     * @param cameraStandardDeviations       the standard deviations of the camera
+     * @return the ambiguity of the estimated pose as a {@link Transform2d}
+     */
+    private Transform2d calculatePoseAmbiguityAsTransform(Pose2d estimatedPoseAtObservationTime, StandardDeviations cameraStandardDeviations) {
+        final Transform2d poseEstimateAtObservationTimeToObservationPose = new Transform2d(estimatedPoseAtObservationTime, estimatedPose);
+        final StandardDeviations estimatedPoseStandardDeviations = cameraStandardDeviations.combineStandardDeviations(PoseEstimatorConstants.ODOMETRY_STANDARD_DEVIATIONS);
+        return estimatedPoseStandardDeviations.scaleTransformFromStandardDeviations(poseEstimateAtObservationTimeToObservationPose);
+    }
```

Comment:
this doesn't calculate ambiguity, this calculates the allowed movement or something from the ambiguity...